Automattic/wp-calypso
# Support for imports from wp-calypso-client

### Background

See #44602

This PR prepares our system to support `wp-calypso-client` imports without dropping support for package-relative imports.

Summary of changes:

* [eslint ]Ignore /packages/wpcom.js/webapp/tests/mocha.js. It is the minified source code of mocha and has tons of violations that we are not going to fix it. It speeds up eslint execution.
* [eslint] Accept imports from the root `package.json`, so it finds `wp-calypso-client`.
* Move `./client/webpack` to `./client/webpack-utils` so it doesn't clash with `node_modules/webpack`. 
* [webpack] Alias `wp-calypso-client` to `./client` for faster resolution and to avoid duplicated modules.
* [webpack] Configure existing externals and replacements to also match `wp-calypso-client` when needed.
* [jest] Replace `wp-calypso-client/config` imports with the actual config file.
* [jest] Transpile `wp-calypso-client` if required.

#### Testing instructions

* Run `yarn start` locally and see it works as expected. Alternatively, verify the live branch works in Chrome and IE11.
* Run `yarn build-server` in this branch and master and verify there are no meaningful differences in `./build/bundle.js`
* Run `MINIFY_JS=false NODE_ENV=production CALYPSO_ENV=production  yarn run build-client-evergreen` in this branch and in master and verify there are no meaningful differences in the evergreen build (for example, compare the entrypoint `public/evergreen/entry-main.*`
* Run tests locally with `yarn test-client` and verify all passes. Alternatively just wait for the CircleCI integration.
